apps/emqx_dashboard/test/emqx_dashboard_mfa_SUITE.erl

```diff
@@ -0,0 +1,342 @@
+    Config;
+t_login_no_mfa_setting_no_mfa_token({'end', _Config}) ->
+    ok;
+t_login_no_mfa_setting_no_mfa_token(_Config) ->
+    Body1 =
+        #{
+            <<"username">> => <<"admin1">>,
+            <<"password">> => <<"admin1pass">>
+        },
+    Body2 = Body1#{<<"mfa_token">> => <<"123456">>},
+    BadBody = Body2#{<<"password">> => <<"wrongpassword1">>},
+    ?assertMatch({ok, 200, _}, login(Body1)),
+    ?assertMatch({ok, 200, _}, login(Body2)),
+    {ok, 401, BadPassRsp} = login(BadBody),
+    ok = assert_return_code("BAD_USERNAME_OR_PWD", BadPassRsp),
+    ok.
+
+%% login when there is MFA state initialized
+%% expect to fail with a TOTP secret prompt
+t_login_with_mfa_setting({init, Config}) ->
+    ok = mock_totp(),
+    Config;
+t_login_with_mfa_setting({'end', _Config}) ->
+    ok = unmock_totp();
+t_login_with_mfa_setting(_Config) ->
+    LoginBody =
+        #{
+            <<"username">> => <<"viewer1">>,
+            <<"password">> => <<"viewer1pass">>
+        },
+    BadPwd = LoginBody#{<<"password">> => <<"wrongpassword1">>},
+    BadTotp = LoginBody#{<<"mfa_token">> => <<"233333">>},
+    BadPwdWithTotp = BadPwd#{<<"mfa_token">> => ?GOOD_TOTP},
+    LoginWithTotp = LoginBody#{<<"mfa_token">> => ?GOOD_TOTP},
+    ?assertMatch({ok, 204, _}, enable_mfa(<<"viewer1">>)),
+    {ok, 401, BadPass} = login(BadPwd),
+    ok = assert_return_code("BAD_USERNAME_OR_PWD", BadPass),
+    %% expect to get a hint about missing MFA token
+    ExpectTotpMissingFn = fun(IsWithSecret) ->
+        {ok, 403, Rsp} = login(LoginBody),
+        #{
+            <<"code">> := <<"BAD_MFA_TOKEN">>,
+            <<"message">> :=
+                #{
+                    <<"error">> := <<"missing_mfa_token">>,
+                    <<"mechanism">> := <<"totp">>
+                } = Msg
+        } = json_map(Rsp),
+        %% assert
+        true = (IsWithSecret =:= maps:is_key(<<"secret">>, Msg)),
+        ok
+    end,
+    %% expect to get a hint about bad MFA token
+    ExpectBadTotpFn = fun(Body) ->
+        {ok, 403, Rsp} = login(Body),
+        #{
+            <<"code">> := <<"BAD_MFA_TOKEN">>,
+            <<"message">> :=
+                #{
+                    <<"error">> := <<"bad_mfa_token">>,
+                    <<"mechanism">> := <<"totp">>
+                } = Msg
+        } = json_map(Rsp),
+        %% assert
+        false = maps:is_key(<<"secret">>, Msg),
+        ok
+    end,
+    ok = ExpectTotpMissingFn(true),
+    %% now EMQX is expecting viewer1 to provide a token for the first time
+    %% but login with a bad password should continue to fail
+    {ok, 401, BadPass1} = login(BadPwd),
+    ok = assert_return_code("BAD_USERNAME_OR_PWD", BadPass1),
+    ok = ExpectTotpMissingFn(true),
+    ok = ExpectBadTotpFn(BadTotp),
+    %% secret is still presented after a failed login with bad token
+    ok = ExpectTotpMissingFn(true),
+    %% try to login with good TOTP but bad password
+    %% this should mark the TOTP setup complete,
```

Comment:
Should there be side-effects if the provided password is wrong?  :thinking:

Reply:
TOTP secret is only provided when password was once correct.
The second login attempt providing a valid TOTP code indicates that the authenticator app is set up.
Otherwise the user will get presented with the QR code again.